oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,486 @@
+worrying about collisions, and it offers slightly more useful introspection
+(timestamp and machine) than a random UUID4. The UUID will be encoded as a
+string using Python's default behavior: lowercase and and dash-separated.
+
+`source <https://github.com/cloudevents/spec/blob/master/spec.md#source-1>`_
+-----------------------------------------------------------------------------
+Example: ``"https://discovery.edx.org/events/"``
+
+This is an absolute URI that represents the URL of the server emitting the
+event. The specific URL path may give some useful information about the events
+being emitted. Groups of servers (e.g. behind a load balancer) should choose the
+same URI, and not have an individual one per server. URIs should be intelligible
+to humans, and not something machine generated like AWS machine names.
+
+Message clients should *not* interpret this value and make switching logic based
+on where a message is coming from. The service producing the event should be
+free to change this URI at any time.
+
+
+`specversion <https://github.com/cloudevents/spec/blob/master/spec.md#specversion>`_
+------------------------------------------------------------------------------------
+Always: ``"1.0"``
+
+Mandatory field that refers to the version of CloudEvents. We have to use "1.0"
+to be spec-compliant.
+
+`type <https://github.com/cloudevents/spec/blob/master/spec.md#type>`_
+----------------------------------------------------------------------
+Example: ``"org.openedx.catalog.course.created.v1"``
+
+This is the name of our event. CloudEvents only requires that this be a string
+describing the type of event, but recommends that it be prefixed with a
+reverse-DNS name for namespacing purposes. We will use a period-separated
+hierarchical name comprising with the format ``{Reverse DNS}-{Architecture
+Domain}-{Subject}-{Action}-{Major Version}``.
+
+Reverse DNS
+~~~~~~~~~~~
+Example: ``org.openedx``
+
+In this example, ``org.openedx`` is the generic prefix that would be used for
+all Open edX installs. Events that are strictly edX-specific, like those that
+might interact with internal IT or finance reporting systems, should use
+``org.edx`` as the prefix instead. If in doubt, default to ``org.openedx``.
+
+Domain (from Domain Driven Design)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Example: ``catalog``
+
+The next part of the ``type`` hierarchy is the Domain. Examples of this are:
+
+* catalog
+* content_authoring
+* credentials
+* ecommerce
+* learning
+
+It is the expectation that there are relatively few domains, and that they will
+roughly match deployed services. Domain names should be lower cased and use
+underscores if they are more than one word.
+
+Subject
+~~~~~~~
+Example: ``course``
+
+The name of an entity that the event applies to. Examples might be ``course``,
+``student``, ``enrollment``, ``order``, etc. Subjects may be namespaced, so
+``special_exam.proctored.allowance`` could be a subject.
```

Comment:
@ormsbee So, recording events to the tracking or sending to a BI backend is not the same thing as interservice communication but it does seem that the event naming could/should be consistent.  Do you imagine deprecating the existing event naming schemas, so something like `edx.bi.user.account.registered` would become something like `org.openedx.learning.student.account.created.v1`?  To my mind, it would be good to maintain consistency of event naming schemas.

Reply:
@bryanlandia: Good point, and I'll add a section to the OEP to try to address it. My general philosophy on events is that backwards compatibility trumps naming consistency. A couple possible approaches:

### 1: Namespace and Encapsulate in CloudEvents

We could put all existing events in a namespace like `org.openedx.legacy.{original event name}`, extract one or two relevant fields, and otherwise echo the event as is in the `data` section.

### 2: Create Parallel Exchange(s)

Instead of trying to retrofit the events to match the new conventions, we create one or two new exchanges. We use the existing event type names as routing keys, but the events are transmitted in exactly the same format they are now, and no attempt is made to modify them to meet any of the conventions proposed in this OEP for new events. We might have a couple of extra exchanges, e.g. one for learner tracking events and one for BI stuff.

New events are added with the new conventions, which I hope will improve versioning and usability overall. Some actions trigger both old and new events. Listeners can register on multiple exchanges. While the listener code might have to change, the code that actually processes the response remains exactly the same.

----

My inclination would be option 2 here. What you you think?

Reply:
ugh, sorry @ormsbee overlooked your response here and appreciate your thoughts.  I'll give this some thought and try to get back to you by end of week.